### ISSUE TITLE
Access and resource configuration on Amarel

### ISSUE THREAD

**Issue by jetesdal** [radical-collaboration/facts#334]
@bobkopp, @pkjr002, @AlexReedy 

I used to be able to run FACTS experiments on Amarel using the following resource.yml

```yaml
resource-desc:
  resource: 'rutgers.amarel'
  walltime: 600
  cpus: 2
  gpus: 0
  queue: 'kopp_1'
  project: ''
  schema: 'local'
```

I am now experiencing issues with running anything even the dummy test with this resource setting 

```
$ python3 runFACTS.py experiments/dummy
EnTK session: re.session.amarel1.amarel.rutgers.edu.jt1187.019868.0011
Creating AppManager
Setting up ZMQ queues                                                         ok
AppManager initialized                                                        ok
Validating and assigning resource manager                                     ok
****** STEP: sealevel_step ******
Setting up ZMQ queues                                                        n/a
All components terminated
Traceback (most recent call last):
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/entk/appman/appmanager.py", line 459, in run
    self._rmgr.submit_resource_request()
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/entk/execman/rp/resource_manager.py", line 225, in submit_resource_request
    self._pilot.wait([rp.PMGR_ACTIVE, rp.DONE, rp.FAILED, rp.CANCELED])
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/pilot/pilot.py", line 585, in wait
    time.sleep(0.1)
KeyboardInterrupt

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/cache/home/jt1187/facts/runFACTS.py", line 193, in <module>
    run_experiment(args.edir, args.debug, args.alt_id, resourcedir=args.resourcedir, makeshellscript = args.shellscript, globalopts = args.global_options)
  File "/cache/home/jt1187/facts/runFACTS.py", line 86, in run_experiment
    amgr.run()
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/entk/appman/appmanager.py", line 485, in run
    raise EnTKError(ex) from ex
radical.entk.exceptions.EnTKError
```

I am able to get the dummy experiment running with the default `local.localhost` resource file
```yaml
resource-desc:
  resource: 'local.localhost'
  walltime: 600
  cpus: 2
  gpus: 0
  queue: ''
  project: ''
  ```
So I assume something has changes such that I can't run FACTS using the Amarel cluster (using the `kopp_1` queue), which previously worked. I am reaching out to clarify whether I still have access to this queue and if there have been any changes to its configuration.

**Comment by bobkopp** [comment]
@AlexReedy can advise. You can no longer launch FACTS from the login node, so if you are doing this it may be the source of your problem.

**Comment by AlexReedy** [comment]
Hey @jetesdal , yes, it seems both the login node and srun capability has been removed from Amarel/RCT to run now you have to run via sbatch script.

Please replace the contents of resource_amarel.yml with this:
# Note the default and requested minimum walltime for RCT is 30 min per RCT user guide
resource-desc:
  resource: 'rutgers.amarel'
  walltime: 7000
  #memory: 128000
  access_schema: 'interactive'
  cpus: 12
  gpus: 0
  queue: 'kopp_1'
  #project: ''
  schema: 'local'

and create a run script in the main facts directory with the following contents
#!/bin/bash
#SBATCH --partition=kopp_1        # Partition (job queue)
#SBATCH --job-name=facts-2node    # Assign a short name to your job
#SBATCH --nodes=1                 # Number of nodes you require
#SBATCH --ntasks-per-node=24      # Total # of tasks across all nodes
#SBATCH --time=05:00:00           # Total run time limit (HH:MM:SS)
#SBATCH --output=slurm.%N.%j.out  # STDOUT output file
#SBATCH --error=slurm.%N.%j.err   # STDERR output file (optional)

source /scratch/ar2208/facts/virtenv/update/bin/activate

radical-stack

export RADICAL_LOG_LVL="DEBUG"
export RADICAL_PROFILE="TRUE"

python3 /scratch/ar2208/facts/branches/fair2/runFACTS.py /scratch/ar2208/facts/branches/fair2/$1

!! NOTE !! you will have to change these paths of course

to run do 
sbatch <name of sbatch script> experiments/<experiment name>

This will create two new files called slurm.<nodename>.<other stuff>, one that is .err and one that is .out

then to get outputs, do
tail -f <name of slurm job>.err

**Comment by jetesdal** [comment]
Thank you for clarifying! I am also using an interactive node for the test. I've been starting from scratch to see if that would fix it, in case the Ensemble Toolkit is updated. Now I am hitting another snag. It's driving me crazy. Here are the steps I am taking to just get it to run:

1. [Insert detailed step-by-step list here]

WHY is it so difficult to get this to work????

**Comment by jetesdal** [comment]
Thank you @bobkopp and @AlexReedy for the suggestions so far. I now have `resources/resource_amarel.yml` as: 
```yml
# Note the default and requested minimum walltime for RCT is 30 min per RCT user guide
resource-desc:
  resource: 'rutgers.amarel'
  walltime: 7000
  #memory: 128000
  access_schema: 'interactive'
  cpus: 12
  gpus: 0
  queue: 'kopp_1'
  #project: ''
  schema: 'local'
```

and have a runscript in the facts directory that looks like this
```bash
#!/bin/bash
#SBATCH --partition=kopp_1        # Partition (job queue)
#SBATCH --job-name=facts-2node    # Assign a short name to your job
#SBATCH --nodes=1                 # Number of nodes you require
#SBATCH --ntasks-per-node=24      # Total # of tasks across all nodes
#SBATCH --time=05:00:00           # Total run time limit (HH:MM:SS)
#SBATCH --output=slurm.%N.%j.out  # STDOUT output file
#SBATCH --error=slurm.%N.%j.err   # STDERR output file (optional)

source /home/jt1187/ve3/bin/activate

radical-stack

export RADICAL_LOG_LVL="DEBUG"
export RADICAL_PROFILE="TRUE"

python3 /home/jt1187/facts/runFACTS.py --global_options '{"rcfg-name": "amarel" }' $1
```

For now I am simply trying to run the dummy experiment using the following command:
```
sbatch /home/jt1187/facts/runscript.sh experiments/dummy
```

I assume I can run the `sbatch` command on a head note. The job immediately fails and I see the following in the `.err` file:
```
Traceback (most recent call last):
  File "/cache/home/jt1187/ve3/bin/radical-stack", line 5, in <module>
    import radical.utils as ru
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/utils/__init__.py", line 29, in <module>
    from .threads        import is_main_thread, is_this_thread, cancel_main_thread
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/utils/threads.py", line 10, in <module>
    import ctypes
  File "/opt/sw/packages/gcc-4_8/python/3.9.6/lib/python3.9/ctypes/__init__.py", line 8, in <module>
    from _ctypes import Union, Structure, Array
ModuleNotFoundError: No module named '_ctypes'
Traceback (most recent call last):
  File "/home/jt1187/facts/runFACTS.py", line 9, in <module>
    import FACTS as facts
  File "/cache/home/jt1187/facts/FACTS.py", line 6, in <module>
    from radical.entk import Pipeline, Stage, Task
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/entk/__init__.py", line 4, in <module>
    from .appman   import AppManager
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/entk/appman/__init__.py", line 2, in <module>
    from .appmanager import AppManager
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/entk/appman/appmanager.py", line 13, in <module>
    import radical.utils   as ru
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/utils/__init__.py", line 29, in <module>
    from .threads        import is_main_thread, is_this_thread, cancel_main_thread
  File "/cache/home/jt1187/ve3/lib/python3.9/site-packages/radical/utils/threads.py", line 10, in <module>
    import ctypes
  File "/opt/sw/packages/gcc-4_8/python/3.9.6/lib/python3.9/ctypes/__init__.py", line 8, in <module>
    from _ctypes import Union, Structure, Array
ModuleNotFoundError: No module named '_ctypes'
```
I  fear that this is now a new error which has to do with the python environment. The issue with the `_ctypes` module not being found seems to be linked to the `radical/entk` package. I have tried to debug this for a few hours now, but could not come up with a solution. The `facts_requirements.txt` specifies `radical.entk==1.42`, but the default installation is 1.60. When I try to force the installation to 1.42, I encounter a conflict error.

Here is the steps when starting from scratch using the [FACTS Quickstart guide](https://fact-sealevel.readthedocs.io/en/latest/quickstart.html).

(1) I cloned the latest version from Github:
```
git clone https://github.com/radical-collaboration/facts.git
```

Instead of downloading the data, I create soft-links to the existing `modules-data` in `/projects/kopp/`
```
cd ~/facts/modules-data
ln -s /projects/kopp/facts-dev/modules-data/*.tgz .
```

Following the guidance from [issue #291](https://github.com/radical-collaboration/facts/issues/291), I am making sure to use python 3.9.6 with
```
module load python/3.9.6
```

And then I followed the exact steps outlined in the [FACTS Quickstart guide](https://fact-sealevel.readthedocs.io/en/latest/quickstart.html) to set up my virtual environment.
```
python3 -m venv ve3
. ve3/bin/activate
pip install --upgrade setuptools pip wheel
pip install radical.entk pyyaml
```

This issue might be related to the system’s configuration or missing system-level dependencies that are typically required for the _ctypes module. I suspect that this is because the installed radical tools are too new to run with FACTS, but I am not able to install the right version. Do you have a shared virtual environment that I can use to avoid these configuration issues? Or could you provide guidance to ensure that the virtual environment meets all the necessary requirements for FACTS to work?

**Comment by andre-merzky** [comment]
@jetesdal : I also had the missing ctypes error on Amarel and opened a support ticket about it.  I did not get any reply so far, I'll ping again.  That error seems unrelated to the RCT stack as I also encountered it on other Python modules...

**Comment by andre-merzky** [comment]
@AlexReedy : the ctype error hits before the RCT stack is even installed.  This is from my Rutgers ticket (still no reply - I have no idea what typical reply times are at Rutgers, do you?):


```sh
[am3465@amarel2 ~]$ module load python/3.8.2

[am3465@amarel2 ~]$ which python3
/opt/sw/packages/gcc-4_8/python/3.8.2/bin/python3

[am3465@amarel2 ~]$ python3 -V
Python 3.8.2

[am3465@amarel2 ~]$ echo $PYTHONPATH

[am3465@amarel2 ~]$ python3 -c 'import ctypes'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/sw/packages/gcc-4_8/python/3.8.2/lib/python3.8/ctypes/__init__.py", line 7, in <module>
    from _ctypes import Union, Structure, Array
ModuleNotFoundError: No module named '_ctypes'
```

**Comment by AlexReedy** [comment]
also this one should work as well to use to most recent version of the RCT stack, I have been sticking with the 1.52 and 1.51 versions but both seem to be working fine
[facts_requirements.txt](https://github.com/radical-collaboration/facts/files/15473154/facts_requirements.txt)

**Comment by jetesdal** [comment]
Thanks @AlexReedy, I tried rebuilding my virtual environment using the provided `facts_requirements.txt file`. However, I encountered the following error:
```
Building wheels for collected packages: radical.entk, radical.gtod, radical.pilot, radical.saga, radical.utils
  Building wheel for radical.entk (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [34 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-t0nmpv_y/radical-entk_7d972b030cd14d28814b30a1022e4938/setup.py", line 255, in <module>
          setup(**setup_args)
        File "/home/jt1187/ve3/lib/python3.9/site-packages/setuptools/__init__.py", line 103, in setup
          return distutils.core.setup(**attrs)
        File "/home/jt1187/ve3/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 171, in setup
          ok = dist.parse_command_line()
        File "/home/jt1187/ve3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 476, in parse_command_line
          args = self._parse_command_opts(parser, args)
        File "/home/jt1187/ve3/lib/python3.9/site-packages/setuptools/dist.py", line 870, in _parse_command_opts
          nargs = _Distribution._parse_command_opts(self, parser, args)
        File "/home/jt1187/ve3/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 535, in _parse_command_opts
          cmd_class = self.get_command_class(command)
        File "/home/jt1187/ve3/lib/python3.9/site-packages/setuptools/dist.py", line 715, in get_command_class
          self.cmdclass[command] = cmdclass = ep.load()
        File "/home/jt1187/ve3/lib/python3.9/site-packages/setuptools/_vendor/importlib_metadata/__init__.py", line 208, in load
          module = import_module(match.group('module'))
        File "/opt/sw/packages/gcc-4_8/python/3.9.6/lib/python3.9/importlib/__init__.py", line 127, in import_module
          return _bootstrap._gcd_import(name[level:], package, level)
        File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
        File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
        File "<frozen importlib._bootstrap_external>", line 850, in exec_module
        File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
        File "/home/jt1187/ve3/lib/python3.9/site-packages/wheel/bdist_wheel.py", line 27, in <module>
          from .macosx_libfile import calculate_macosx_platform_tag
        File "/home/jt1187/ve3/lib/python3.9/site-packages/wheel/macosx_libfile.py", line 43, in <module>
          import ctypes
        File "/opt/sw/packages/gcc-4_8/python/3.9.6/lib/python3.9/ctypes/__init__.py", line 8, in <module>
          from _ctypes import Union, Structure, Array
      ModuleNotFoundError: No module named '_ctypes'
      [end of output]
```

The previous version I shared was installed by forcing the version of entk to be 1.42. This is because I saw this version in the [facts_requirments.txt](https://github.com/radical-collaboration/facts/blob/main/facts_requirments.txt) uploaded onto the repo. Using `facts_requirments.txt` that you shared (with entk version 1.52.0) causes the installation to fail.

**Comment by AlexReedy** [comment]
yea sorry I have not updated the main branch version yet, I'll do that now.

Also did you run pip install --upgrade setuptools pip wheel before running pip install -r facts_requirements.txt?

**Comment by jetesdal** [comment]
I think something on the Amarel system is out of whack and that causes the installation to fail. The idea I had is to sjip the installation step by using a common envoronemt my just using something like
`source /projects/kopp/facts/virtenv/bin/activate`

**Comment by AlexReedy** [comment]
can you go step by step how you are installing your ve? please be as specific as possible including info on the base conda env

**Comment by jetesdal** [comment]
Sure, here is what I do. After logging into a head note I usually have a conda env:

```
(base) [jt1187@amarel2 ~]$ 
(base) [jt1187@amarel2 ~]$ which python3
~/miniconda3/bin/python3
```

For now I deactivate the conda environment and load the community Python 3.9.6 (based on @bobkopp comments in #291)
```
conda deactivate
module load python/3.9.6
```

So I now have 
```
$ which python3
/opt/sw/packages/gcc-4_8/python/3.9.6/bin/python3
```

To start from scratch I remove the previous environment
```
rm -rf /home/jt1187/ve3
```
and created a new one as shown in [quickstart](https://fact-sealevel.readthedocs.io/en/latest/quickstart.html):
```
python3 -m venv ve3
. ve3/bin/activate
```

Upgraded pip, setuptools, and wheel:
```
pip install --upgrade setuptools pip wheel
```

Installed required packages with provided .txt file
```
pip install -r /home/jt1187/facts_requirements.txt
```

**Comment by andre-merzky** [comment]
> also this one should work as well to use to most recent version of the RCT stack, I have been sticking with the 1.52 and 1.51 versions but both seem to be working fine [facts_requirements.txt](https://github.com/radical-collaboration/facts/files/15473154/facts_requirements.txt)

As long as those versions are working I see little need to update the stack actually.  Never change a running system, right? ;-)

**Comment by AlexReedy** [comment]
@andre-merzky yea I agree! haha

@jetesdal try instead of loading the community module, building the python ve with the base conda 3.9.6 activate, _then_ deactivate it. It's a bit wonkly but I started doing it that way to use python 3.11 a while ago and it's the method I have been using on Amarel since. So try this:
1. with (base) active: $python3 -m venv <ve name>
2. conda deactivate
3. source <ve name>/bin/activate
4. $pip install --upgrade setuptools pip wheel
5. pip install -r facts_requirements.txt

**Comment by jetesdal** [comment]
I have an update on the `_ctypes` error. I found a solution that works for me. I ended up using Miniforge to avoid any inheritance of system libraries.

```bash
wget https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh
bash Miniforge3-Linux-x86_64.sh
```
With the activated  (base) environment, I then followed @AlexReedy's recommendations  to create the virtual environment:
```bash
python3 -m venv ve3
. ve3/bin/activate
pip install --upgrade setuptools pip wheel
pip install radical.entk pyyaml
```
I think the trick is to avoid using the community Python model. I also heard back from help@oarc.rutgers.edu. I ended up not trying out their suggestions, but might be useful to know about:
> You can use the existing libffi in the community modules and see if the error goes away.
> module use /projects/community/modulefiles
> module avail libffi
> 
> You can also use the existing Python 3.9 version by
> module load python/3.9.6-gc563

**Comment by jetesdal** [comment]
Now that the environment seems to be setup right and there is no more missing `_ctypes` module issue when running the dummy test, I proceeded with the test run as described in the [quickstart](https://fact-sealevel.readthedocs.io/en/latest/quickstart.html):
```bash
cd ~/facts/
mkdir test
cp -r experiments/coupling.ssp585/config.yml test
sbatch /home/jt1187/facts/runscript.sh test
```

I am using the same `resources/resource_amarel.yml` and the `runscript.sh` with 
```bash
python3 /home/jt1187/facts/runFACTS.py --global_options '{"rcfg-name": "amarel" }' $1
```

It has run until the time out:
```vbnet
slurmstepd: error: *** JOB 36893742 ON hal0262 CANCELLED AT 2024-05-29T21:38:48 DUE TO TIME LIMIT ***
```
I attached the `.err` file here:
[slurm.hal0262.36893742.err.txt](https://github.com/radical-collaboration/facts/files/15492783/slurm.hal0262.36893742.err.txt)

I also found several task failures:
```vbnet
state: FAILED
task task.000032 failed: RuntimeError("task failed")
```

 I am not sure if they are related to the time out. It seems odd that this simple test run (using a single location) exceeds the allocated time and gets canceled. @AlexReedy do you recommend increasing the time limit or is there something I'm missing? I am planning to look into the sandbox, but I suspect that it has to do with either bad resource allocation or general issue with Amarel system. Any suggestions on how to address this would be greatly appreciated!

**Comment by andre-merzky** [comment]
Looking into the sandbox is likely the best course of action to determine what failed and why.  Feel free to attach the pilot sandbox to this ticket of you want me to have a look.

**Comment by AlexReedy** [comment]
No need @andre-merzky it looks like it's just a runtime error. @jetesdal try changing the time in here

#!/bin/bash
#SBATCH --partition=kopp_1        # Partition (job queue)
#SBATCH --job-name=facts-2node    # Assign a short name to your job
#SBATCH --nodes=1                 # Number of nodes you require
#SBATCH --ntasks-per-node=24      # Total # of tasks across all nodes
#SBATCH --time=05:00:00           # Total run time limit (HH:MM:SS)
#SBATCH --output=slurm.%N.%j.out  # STDOUT output file
#SBATCH --error=slurm.%N.%j.err   # STDERR output file (optional)

source /home/jt1187/ve3/bin/activate

radical-stack

export RADICAL_LOG_LVL="DEBUG"
export RADICAL_PROFILE="TRUE"

python3 /home/jt1187/facts/runFACTS.py --global_options '{"rcfg-name": "amarel" }' $1

-------------------------------

and in here under wall time:

# Note the default and requested minimum walltime for RCT is 30 min per RCT user guide
resource-desc:
  resource: 'rutgers.amarel'
  walltime: 7000
  #memory: 128000
  access_schema: 'interactive'
  cpus: 12
  gpus: 0
  queue: 'kopp_1'
  #project: ''
  schema: 'local'
--------------------------------
!! NOT !! make sure they match and I think amarel allows up to 13 or 14 days or something. Also the wall time parameter in the resource.yml file is in minutes

**Comment by jetesdal** [comment]
> Looking into the sandbox is likely the best course of action to determine what failed and why. Feel free to attach the pilot sandbox to this ticket of you want me to have a look.

@andre-merzky, can you clarify what the pilot sandbox is? Is it the directly in my `facts/` path that is created? This would be `re.session.hal0262.amarel.rutgers.edu.jt1187.019872.0003`

**Comment by jetesdal** [comment]
There is also `/scratch/jt1187/radical.pilot.sandbox/re.session.hal0262.amarel.rutgers.edu.jt1187.019872.0003/`
but that one is 31 GB large. Not sure how to attach this here.

**Comment by jetesdal** [comment]
> @jetesdal try changing the time in here and yea send over the sandbox as well. But did it run for 5 hours and then fail?

I am not sure. It run for at least a couple of hours, more than I would have expected for a single location. I will go ahead and use your suggestion to increase the time.

Just to clarify, so if I set the time in `runscript.sh` to say `08:00:00`, I should set the walltime in `resource_amarel.yml` to `28800` (8h x 60min x 60s )? Is the same true for the number of nodes (e.g., `--nodes=2 `) or the `--ntasks-per-node`? For example, if I have `--ntasks-per-node=16`  in `runscript.sh` I also should have `cpus: 16` in `resource_amarel.yml`?

**Comment by AlexReedy** [comment]
What experiment are you running?

**Comment by jetesdal** [comment]
Hi @AlexReedy, I am just trying the example test run as described in the [quickstart](https://fact-sealevel.readthedocs.io/en/latest/quickstart.html):

```
cd ~/facts/
mkdir test
cp -r experiments/coupling.ssp585/config.yml test
sbatch /home/jt1187/facts/runscript.sh test
```

The config file is attached: [config.yml.txt](https://github.com/radical-collaboration/facts/files/15503993/config.yml.txt)

**Comment by AlexReedy** [comment]
how many locations?

**Comment by jetesdal** [comment]
Yes, just one location. Since I don't provide a location file it uses, New York (id: 12):
https://github.com/radical-collaboration/facts/blob/9fafbfcad6994fcdd1bb3afc137612872493ab05/runFACTS.py#L31-L34

I use the latest version of facts main branch:
```
git clone https://github.com/radical-collaboration/facts.git
```

**Comment by AlexReedy** [comment]
hmm ok, so it sounds like it's hanging... can you try from this branch and let me know if you're getting he same result?
https://github.com/radical-collaboration/facts/tree/facts/memory
also do you know _where_ it's hanging?

**Comment by jetesdal** [comment]
@AlexReedy, I followed your advice and used the `facts/memory` branch. That made the test run complete successfully without any time outs. I also was able to get the test run completed with no time outs on the main branch, after increasing the time limit and ensuring consistency between `runscript.sh` and the `resource_amarel.yml`.

I'm now confident that I have FACTS working on Amarel again, and I think we can close this issue. Thank you very much for your suggestions and support!
